src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonIDictionaryConverter.cs

```diff
@@ -0,0 +1,113 @@
+        : JsonDictionaryDefaultConverter<TCollection, object>
+        where TCollection : IDictionary
+    {
+        protected override void Add(object value, JsonSerializerOptions options, ref ReadStack state)
+        {
+            string key = state.Current.JsonPropertyNameAsString!;
+            ((IDictionary)state.Current.ReturnValue!)[key] = value;
+        }
+
+        protected override void CreateCollection(ref ReadStack state)
+        {
+            JsonClassInfo classInfo = state.Current.JsonClassInfo;
+
+            if (TypeToConvert.IsInterface || TypeToConvert.IsAbstract)
+            {
+                if (!TypeToConvert.IsAssignableFrom(RuntimeType))
+                {
+                    ThrowHelper.ThrowNotSupportedException_DeserializeNoParameterlessConstructor(TypeToConvert);
+                }
+
+                state.Current.ReturnValue = new Dictionary<string, object>();
+            }
+            else
+            {
+                if (classInfo.CreateObject == null)
+                {
+                    ThrowHelper.ThrowNotSupportedException_DeserializeNoParameterlessConstructor(TypeToConvert);
+                }
+                else
+                {
+                    TCollection returnValue = (TCollection)classInfo.CreateObject!()!;
+
+                    if (returnValue.IsReadOnly)
+                    {
+                        ThrowHelper.ThrowNotSupportedException_SerializationNotSupported(TypeToConvert);
+                    }
+
+                    state.Current.ReturnValue = returnValue;
+                }
+            }
+        }
+
+        protected internal override bool OnWriteResume(Utf8JsonWriter writer, TCollection value, JsonSerializerOptions options, ref WriteStack state)
+        {
+            IDictionaryEnumerator enumerator;
+            if (state.Current.CollectionEnumerator == null)
+            {
+                enumerator = value.GetEnumerator();
+                if (!enumerator.MoveNext())
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                enumerator = (IDictionaryEnumerator)state.Current.CollectionEnumerator;
+            }
+
+            JsonConverter<object> converter = GetValueConverter(ref state);
+            do
+            {
+                if (!(enumerator.Key is string key))
+                {
+                    // todo: add test for this.
+                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(state.Current.DeclaredJsonPropertyInfo.RuntimePropertyType!);
+                    Diagnostics.Debug.Assert(false);
+                    return false;
+                }
+
+                key = GetKeyName(key, ref state, options);
+                writer.WritePropertyName(key);
+
+                object? element = enumerator.Value;
+                if (!converter.TryWrite(writer, element!, options, ref state))
```

Comment:
This is a special case with generics. TryWrite is a generic method for `element` which cannot be qualified as struct or class since it can either, so it can't have ? applied:
```error CS8627: A nullable type parameter must be known to be a value type or non-nullable reference type. Consider adding a 'class', 'struct', or type constraint```

The compiler error will be generated if we change the class to `JsonConverter<object?>`

PLMK if I'm missing something.

Reply:
I am **not** suggesting we change the signature of the `TryWrite` generic method at all.

Let the null-ability of the generic `T` drive that in the `TryWrite` method.

This change alone (just to the `JsonIDictionaryConverter` file) compiles fine for me:
![image](https://user-images.githubusercontent.com/6527137/74070912-3083ac80-49b7-11ea-8de9-eab18901d6f2.png)

Reply:
I see. Yep these can be nullable.